ext/surro-gate/proxy.c

```diff
@@ -0,0 +1,185 @@
+    // Rearm right socket for writing and also reading if not ready for reading
+    selector_rearm(selector, SOCK_PTR(rb_iv_get(pair, "@right")), NUM2INT(inv_idx), NUM2INT(argv[1]), EPOLLOUT | (IVAR_TRUE(inverse, "@rd_rdy") ? 0 : EPOLLIN));
+  }
+  return Qnil;
+}
+
+static VALUE SurroGate_Proxy_allocate(VALUE self) {
+  int *selector = malloc(sizeof(int));
+
+  if (selector != NULL) {
+    *selector = selector_init();
+    if (*selector > 0) {
+      return Data_Wrap_Struct(self, NULL, SurroGate_Proxy_free, selector);
+    } else {
+      xfree(selector);
+    }
+  }
+
+  rb_raise(rb_eRuntimeError, "Allocation failed!");
+  return Qnil;
+}
+
+static void SurroGate_Proxy_free(int *selector) {
+  close(*selector);
+  xfree(selector);
+}
+
+static VALUE SurroGate_Proxy_initialize(VALUE self, VALUE logger) {
+  VALUE mConcurrent = rb_const_get(rb_cObject, rb_intern("Concurrent"));
+  VALUE cArray = rb_const_get(mConcurrent, rb_intern("Array"));
+
+  rb_iv_set(self, "@pairing", rb_class_new_instance(0, NULL, cArray)); // @pairing = Concurrent::Array.new
+  rb_iv_set(self, "@logger", logger);
+
+  return Qnil;
+}
+
+static VALUE SurroGate_Proxy_push(VALUE self, VALUE left, VALUE right) {
+  int index_ltr, index_rtl, *selector;
+
+  VALUE pair_LTR[2] = {left, right};
+  VALUE pair_RTL[2] = {right, left};
+  VALUE left_to_right = rb_class_new_instance(2, pair_RTL, cSurroGate_Pair); // SurroGate::Pair.new(left, right)
+  VALUE right_to_left = rb_class_new_instance(2, pair_LTR, cSurroGate_Pair); // SurroGate::Pair.new(right, left)
+
+  VALUE pairing = rb_iv_get(self, "@pairing");
+
+  Check_Type(left, T_FILE);
+  Check_Type(right, T_FILE);
+
+  // raise ArgumentError if @pairing.detect(&pairing_compare)
+  if (rb_block_call(pairing, rb_intern("detect"), 0, NULL, pairing_compare, rb_ary_new_from_values(2, pair_LTR)) != Qnil) {
+    rb_raise(rb_eArgError, "Socket already registered!");
+  }
+
+  Data_Get_Struct(self, int, selector);
+
+  rb_funcall(pairing, rb_intern("push"), 2, left_to_right, right_to_left); // @pairing.push(left_to_right, right_to_left)
+  index_ltr = NUM2INT(rb_funcall(pairing, rb_intern("index"), 1, left_to_right)); // @pairing.index(left_to_right)
+  index_rtl = NUM2INT(rb_funcall(pairing, rb_intern("index"), 1, right_to_left)); // @pairing.index(right_to_left)
+
+  rb_iv_set(left_to_right, "@inverse", INT2NUM(index_rtl));
+  rb_iv_set(right_to_left, "@inverse", INT2NUM(index_ltr));
+
+  selector_register(selector, SOCK_PTR(left), index_ltr, index_rtl);
+  selector_register(selector, SOCK_PTR(right), index_rtl, index_ltr);
+
+  return Qtrue;
+}
+
+static VALUE SurroGate_Proxy_pop(VALUE self, VALUE sockets) {
+  int i, *selector;
+
+  VALUE pairing = rb_iv_get(self, "@pairing");
+  rb_block_call(pairing, rb_intern("delete_if"), 0, NULL, pairing_compare, sockets); // @pairing.delete_if(&pairing_compare)
```

Comment:
Little confused why there may be multiple sockets here.
maybe confused what exactly are sockets - are there 2  - one for each direction?

Reply:
So you have two sockets, let's say `a` and `b`. If you push these into the proxy, you will have 2 pairing objects `a->b` and `b->a`.